src/alacris_lib/state_update.ml

```diff
@@ -1,99 +1,120 @@
-
-type request_state_update =
-  | Submit of (Digest.t * TransactionReceipt.t OrExn.t Lwt.u)
-  | GetLastRevision of Revision.t Lwt.u
-  | GetLastCommit of Digest.t Lwt.u
-
-let request_state_update_mailbox : request_state_update Lwt_mvar.t = Lwt_mvar.create_empty ()
-
-let post_to_mailbox_state_update : Digest.t -> TransactionReceipt.t OrExn.t Lwt.t =
-  fun digest ->
-  simple_client request_state_update_mailbox
-    (fun ((_x_digest, x_resolver) : (Digest.t * TransactionReceipt.t OrExn.t Lwt.u)) -> Submit (digest,x_resolver)) digest
-
-
-let retrieve_last_posted_state : unit -> Digest.t Lwt.t =
-  fun () ->
-  simple_client request_state_update_mailbox
-    (fun ((_x_unit, x_resolv) : (unit * Digest.t Lwt.u)) -> GetLastCommit x_resolv) ()
-
-let retrieve_last_revision : unit -> Revision.t Lwt.t =
-  fun () ->
-  simple_client request_state_update_mailbox
-    (fun ((_x_unit, x_resolv) : (unit * Revision.t Lwt.u)) -> GetLastRevision x_resolv) ()
-
-
-let init_state : unit -> digest_entry =
-  fun () -> {revision = Revision.of_int 0; digest = null_digest}
-
+let state_update_log = true
 
-let the_digest_entry_ref : (digest_entry ref) = ref (init_state ())
+let status_null_operation = ref 0
 
+let last_hash = ref Digest.zero
+let first_nontrivial_hash = ref Digest.zero
 
 
-
-let post_state_update : Revision.t -> Digest.t -> TransactionReceipt.t Lwt_exn.t =
-  fun operator_revision operator_digest ->
+let post_state_update : operator:Address.t -> operator_revision:Revision.t -> operator_digest:Digest.t -> unit Lwt_exn.t =
+  fun ~operator ~operator_revision ~operator_digest ->
+  let open Lwt_exn in
   if state_update_log then
-    Logging.log "post_state_update operator_revision=%s operator_digest=%s" (Revision.to_string operator_revision)  (Digest.to_0x operator_digest);
+    Logging.log "post_state_update operator_revision=%s digest=%s" (Revision.to_string operator_revision) (Digest.to_0x operator_digest);
+  if (String.equal (Digest.to_string !last_hash) (Digest.to_string operator_digest)) then
+    (if state_update_log then
+       Logging.log "Same hash as before. No need to do state_update. Instead doing small_money_transfer";
+     status_null_operation := 1 - !status_null_operation;
+     Side_chain_null_operation.small_money_transfer !status_null_operation
+    )
+  else
+    (if state_update_log then
+       Logging.log "New hash, doing a state_update";
+     last_hash := operator_digest;
+     get_contract_address ()
+     >>= fun contract_address ->
+     let operation = make_state_update_call ~contract_address ~operator_digest ~operator_revision in
+     Ethereum_user.post_operation ~operation:operation ~sender:operator ~value_send:TokenAmount.zero
+     >>= fun _ ->
+     if state_update_log then
+       Logging.log "After the post_operation of post_state_update";
+     return ()
+    )
+
+
+let post_state_update_nocheck : operator:Address.t -> operator_revision:Revision.t -> operator_digest:Digest.t -> unit Lwt_exn.t =
```

Comment:
I renamed it **_for_test**.
The state_update now uses a require statement. As a consequence, we cannot do two claims consecutively which are identical. This is where the needs for having small activity came from.
For the tests, we need to use different binary but the same blockchain. Since the first hash is starting from zero we need for the test not to be doing the first one.
This requires a different a different function for the state update of the test.